rclcpp/doc/api_review_march_2020.md

```diff
@@ -0,0 +1,433 @@
+
+- https://github.com/ros2/rclcpp/issues/979
+
+**Suggested Action**: Continue with issue.
+
+**Notes from 2020-03-23**:
+
+- (shane) says "yes please" :)
+
+### Use `std::variant` instead of custom `ParameterValue` class
+
+> This is only possible if C++17 is available, but it would simplify our code, make our interface more standard, and allow us to use constexpr-if to simply our templated code.
+
+**Suggested Action**: Create an issue for future work.
+
+**Notes from 2020-03-23**:
+
+- (chris) not sure churn is worth
+- (ivan) other places for std::variant, like AnySubscriptionCallback
+
+### Cannot set name or value on `Parameter`/`ParameterValue`
+
+> Both `Parameter` and `ParameterValue` are read-only after construction.
+
+- https://github.com/ros2/rclcpp/issues/238
+
+**Suggested Action**: Update issue, possibly close.
+
+**Notes from 2020-03-23**:
+
+- (chris/william) setting values on temporary (local) objects is not reflected in the node, so misleading
+
+## Parameter Clients
+
+### No timeout option with synchronous parameter client calls
+
+> As an example, SyncParametersClient::set_parameters doesn't take a timeout option. So, if anything goes wrong in the service call (e.g. the server goes down), we will get stuck waiting indefinitely.
+
+- https://github.com/ros2/rclcpp/issues/360
+- https://github.com/ros2/rclcpp/blob/96ebf59a6045a535730d98fff25e522807c7aa75/rclcpp/src/rclcpp/parameter_client.cpp#L453-L468
+
+**Suggested Action**: Update issue, decide if it can be taken for Foxy or not.
+
+**Notes from 2020-03-23**:
+
+- (tfoote) Seems like adding a timeout is a good idea.
+
+### Name of AsyncParametersClient inconsistent
+
+> AsyncParameter**s**Client uses plural, when filename is singular (and ParameterService is singular):
+
+- https://github.com/ros2/rclcpp/blob/7c1721a0b390be8242a6b824489d0bc861f6a0ad/rclcpp/include/rclcpp/parameter_client.hpp#L44
+
+**Suggested Action**: Reconcile class and file name, switch to singular name?
+
+**Notes from on-line, post 2020-03-23 meeting**:
+
+- 
+
+### `SyncParametersClient::get_parameters` doesn't allow you to detect error cases
+
+> E.g. https://github.com/ros2/rclcpp/blob/249b7d80d8f677edcda05052f598de84f4c2181c/rclcpp/src/rclcpp/parameter_client.cpp#L246-L257 return an empty vector if something goes wrong which is also a valid response.
+
+- https://github.com/ros2/rclcpp/issues/200
+- https://github.com/ros2/rclcpp/blob/96ebf59a6045a535730d98fff25e522807c7aa75/rclcpp/src/rclcpp/parameter_client.cpp#L412-L426
+
+**Suggested Action**: ?
+
+**Notes from on-line, post 2020-03-23 meeting**:
+
+- 
+
+## Clock
+
+### Clock Jump callbacks on System or Steady time?
+
+> Currently time jump callbacks are registered via Clock::create_jump_handler(). Jump handlers are only invoked by TimeSource::set_clock(). This is only called if the clock type is RCL_ROS_TIME and ROS time is active.
+
+- https://github.com/ros2/rclcpp/issues/528
+
+**Suggested Action**: ?
```

Comment:
There should be no jumps in steady time. If there's a big change in system time, it doesn't necessarily mean that time jumped, just that you might have been sleeping for a long time. Most ntp systems adjust the slew rate these days instead of jumping but still that's an external process and I don't know of any APIs to introspect the state of the clock.

I'm not sure that we have a way to detect jumps in time for system or steady time. To that end I think that we should be clear that we only provide callbacks when simulation time starts or stops, or simulation time jumps.

We should also strongly recommend that operators not actively adjust their system clocks while running ROS nodes.

Reply:
I agree with Tully, if we don't have a way to detect system time jumps then I think we should just document that this only works with ROS time. In addition to documentation, we could log an info or warning message if the user registers jump callback with steady or system time, but it may be unnecessarily noisy.